### PR TITLE
139 data coverage

### PR DESCRIPTION
**Pull request by loffleraSMH**
Note: Yishan & Danika have already reviewed a previous version of this function. The main purpose of the additional review is to focus on the vignette and some additional enhancements I made to the `data_coverage` functions.

The `data_coverage()` function facilitates data coverage exploration. It returns encounter-level flags based on the `lookup_data_coverage` table (indicating whether a given `genc_id` was discharged during a time period with coverage for a given table). By default it also generates a timeline plot visualizing the min-max dates for each hospital*table combination (see "Data availability" plot in data dictionary) and a coverage plot showing the % `genc_ids` with an entry in a given table (by hospital*month).

- Please start by reviewing the `data_coverage` vignette for additional explanations & use-case scenarios. A knitted version of the vignette can be found here: `/mnt/nfs/projects/research_projects/Rgemini/data_coverage.html`. Please check the vignette for clarity and make sure it contains all relevant information.
- After reviewing the vignette, please do a final review of the `data_coverage` function itself. I have added some additional customization (e.g., `custom_dates`, `hospital_label` and `hospital_group` inputs) after Yishan & Danika reviewed the function. Please make sure that this works as expected and test the function on any edge cases you can think of.

You can use the following code as a starting point for testing:
```
################ Set up ################
library(DBI)
library(RPostgreSQL)
library(getPass)
library(dplyr)
library(data.table)
library(Rgemini)
library(lubridate)
library(ggplot2)
library(lemon)

drv <- dbDriver("PostgreSQL")
dbcon <- DBI::dbConnect(drv,
                    dbname = "drm_cleandb_v3_1_0", # "drm_cleandb_v2_1_3", "drm_cleandb_v3_1_0"
                    host = "prime.smh.gemini-hpc.ca",
                    port = 5432,
                    user = getPass("User:"),
                    password = getPass("Enter Password:"))

cohort <- dbGetQuery(dbcon, "SELECT genc_id, discharge_date_time, hospital_id FROM admdad WHERE discharge_date_time >= '2017-01-01 00:00';") %>% data.table()

# source relevant updated functions (adjust path according to your Rgemini folder)
source("~/GitHub/Rgemini/R/data_coverage.R")
source("~/GitHub/Rgemini/R/utils.R") 
source("~/GitHub/Rgemini/R/plot_over_time.R") 

################ Run function without plots ################
## Get data coverage lookup table with flag by genc_id
coverage <- data_coverage(
    dbcon, cohort, c("admdad", "physicians"),
    plot_timeline = FALSE,
    plot_coverage = FALSE,
    as_plotly = FALSE
)
head(coverage$coverage_flag_enc)


################ Run function with plot_timeline = TRUE #####################
## Plot min-max available dates ("Data Timeline")
coverage <- data_coverage(
    dbcon, cohort,
    c("admdad", "ipscu", "transfusion"),
    plot_timeline = TRUE,
    plot_coverage = FALSE
)
## this returns a list with 2 items
# the 1st one (coverage_flag_enc) is the same as above (lookup table with flag for each genc_id)
head(coverage[[1]]) # = coverage$coverage_flag_enc
# the 2nd one (timeline_data) is the DB availability table with min-max dates per site & table (i.e., what's shown in the data timeline plot)
head(coverage[[2]]) # = coverage$timeline_data


################ Run function with plot_coverage = TRUE #####################
coverage <- data_coverage(
    dbcon, cohort,
    "radiology",
    plot_timeline = FALSE,
    plot_coverage = TRUE,
    colors = "black"
)
## this returns a list with 2 items
head(coverage[[1]]) # encounter-level coverage flag (see above)
# the 2nd list item shows the number of encounters per month, and % of those encounters with an entry in the respective table
head(coverage[[2]]) # coverage$coverage_data


################ Plot everything (default) with additional customization #####################
# add hospital type
hosp_lookup <- dbGetQuery(dbcon, "SELECT * FROM lookup_hospital;") %>% data.table()
cohort <- merge(cohort, hosp_lookup[, .(hospital_id, hospital_type)], by = "hospital_id", all.x = TRUE)

# add numeric hospital label from 1-32
set(cohort, j = "hospital", value = factor(cohort$hospital_id, levels = unique(sort(cohort$hospital_id)), labels = 1:length(unique(sort(cohort$hospital_id)))))

cohort[, .(hospital_id, hospital, hospital_type)] %>%
    distinct()

# define custom dates
custom_dates <- data.table(
    hospital_id = "HRH",
    data = "clinicalnotes",
    min_date = as.Date("2020-01-01"),
    max_date = as.Date("2022-06-01")
)

# create plots
coverage <- data_coverage(
    dbcon, cohort,
    "clinicalnotes",
    hospital_label = "hospital",
    hospital_group = "hospital_type",
    custom_dates = custom_dates,
    colors = gemini_colors(3)#,
    #as_plotly = TRUE
)
```

### CODE REVIEW

**Comment by danikanou** [comment]
I think the vignette looks great! Will be testing the function now.

File: R/data_coverage.R [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+      stop(paste(
+        "Please make sure the `hospital_group` variable is a hospital",
+        "grouping variable that has a 1:1 relationship with the",
+        "hospital identifier (e.g., hospital_num)."
+      ))
+    }
+  }
+
+  # check that custom_dates has correct format
+  if (!is.null(custom_dates)) {
```
Comment (danikanou) [inline R/data_coverage.R:254]: custom dates fail to work when they are provided in a format other than yyyy-mm-dd (even yyyy/mm/dd) as they are considered ambiguous by as.Date. Maybe we should check the format or use a function convert_dt from Rgemini or Lubridate's parse_date_time to clean up the provided format.

Comment (loffleraSMH) [reply R/data_coverage.R:254]: Great catch! I actually decided to enforce ymd formats to avoid ambiguity (and because that's consistent with other `Rgemini` functions/`drm_cleandb`), but I added `convert_dt` to allow for a bit more flexibility with that, so yyyy/mm/dd should work now. I also clarified that in the documentation and added a more informative error message when the formats fail. 

I also added `hospital_label` (if provided) to the message showing hospital-level information. I had to make a couple of other changes for that throughout the code, so it's worth double checking that I didn't break anything. I ran some of my own tests and everything seems to work fine, but please confirm to be sure. Thank you!

Comment (danikanou) [reply R/data_coverage.R:254]: It works well now! Just one more thing, when I test the function without forcing as.Date on my minimum and maximum dates when they are ambiguous like so: `custom_dates <- data.table(
    hospital_id = "HRH",
    data = "clinicalnotes",
    min_date = "01012021",
    max_date = "01012022"
)
`, the function is interpreting the min_date as "2001-01-20"- which makes sense because it's ambiguous. I found that `as.Date(convert_dt(as.Date("01012021"), c("ymdHMS"), truncated = 3))` does catch the ambiguous format and throws an error- but it's a very wordy line... Maybe we could enforce as.Date on the min and max dates earlier on in the function?

**Review by danikanou — COMMENTED** [review]
I made a comment about a very minor issue with ambiguous date formats being misinterpreted, but otherwise the function works really well!

File: R/data_coverage.R (hunk 1 above)
Comment (loffleraSMH) [reply R/data_coverage.R:254]: Hm, the problem with using `as.Date(convert_dt(as.Date(...)))` is that the function then again doesn't handle some of the formats correctly even when they are in yyyy-mm-dd, like `as.Date(convert_dt(as.Date("20211201"), c("ymdHMS"), truncated = 3))`. So I think to avoid confusion, we should enforce the yyyy-mm-dd format and then let `convert_dt` take priority over `as.Date()`. Users should notice if they provided the wrong dates based on the timeline plot anyway. 

In terms of the error message we show, do you mean this one? `"Invalid format for min/max dates in 'custom_dates' input. Please make sure dates are specified in 'yyyy-mm-dd' format (e.g., min_date = '2020-12-01')."` Let me know if you want me to rephrase/shorten that.

I also just fixed a small typo in the roxygen (the threshold for a gap in coverage is 28 days, not 30 days).

Comment (danikanou) [reply R/data_coverage.R:254]: Right, that sounds good! Sorry, don't worry about the error.

I had a look at the recent commits and the roxygen edits as well, and they look good!

**Comment by danikanou** [comment]
I finished reviewing + testing, and I think that the function is ready for push. Thanks for your hard work Anne!

**Comment by loffleraSMH** [comment]
> I finished reviewing + testing, and I think that the function is ready for push. Thanks for your hard work Anne!

Awesome, thanks so much for the thorough review @danikanou! I'll go ahead and merge this :-)
